server/pg/serialize.cpp

```diff
@@ -234,6 +235,90 @@ void SerializeVarchar(SerializationContext context,
   }
 }
 
+template<VarFormat Format, typename UnscaledType>
+void SerializeDecimal(SerializationContext context,
+                      const velox::DecodedVector& decoded_vector,
+                      velox::vector_size_t row) {
+  const auto& type = decoded_vector.base()->type();
+  const auto [precision, scale] = velox::getDecimalPrecisionScale(*type);
+  auto value = decoded_vector.valueAt<UnscaledType>(row);
+  if constexpr (Format == VarFormat::Text) {
+    size_t max_size = velox::DecimalUtil::maxStringViewSize(precision, scale);
+    context.buffer->WriteContiguousData(max_size, [&](auto* data) {
+      char* buf = reinterpret_cast<char*>(data);
+      return velox::DecimalUtil::castToString<UnscaledType>(
+        value, scale, static_cast<int32_t>(max_size), buf);
+    });
+  } else {
+    // Well, here we go...
+    // Postgre numeric(decimal) type has special binary layout:
+    // int16_t ndigits;  number of digits
+    // int16_t weight;  weight of first digit(the exponent of the first digit)
+    // int16_t sign;  sign of the number
+    // int16_t dscale;  display scale
+    // int16_t digits[ndigits];  base 10000 digits
+
+    // So, in velox value = real_value * 10^scale
+    // However, in Postgres each digit represents 4 decimal digits,
+    // i.e. base 10000. So we need to convert velox decimal representation
+    // to Postgres numeric representation.
+
+    // E.g. value = 12345 with scale = 2 means real_value = 123.45
+    // In Postgres representation it will be:
+    // ndigits = 2
+    // weight = 0
+    // sign = 0x0000
+    // dscale = 2
+    // digits[0] = 123
+    // digits[1] = 4500
+    constexpr size_t kBaseSystem = 10'000;
+    constexpr int16_t kPositive = 0x0000;
+    constexpr int16_t kNegative = 0x4000;
+    int16_t ndigits = 0;
+    {
+      auto val = value;
+      while (val != 0) {
+        val /= kBaseSystem;
+        ++ndigits;
+      }
+    }
+    ndigits = scale % 4 == 0 ? ndigits : ndigits + 1;
+    int16_t weight = static_cast<int16_t>(ndigits - ((scale + 3) / 4) - 1);
+    int16_t sign = (value < 0) ? kNegative : kPositive;
+    int16_t dscale = static_cast<int16_t>(scale);
+    value = value < 0 ? -value : value;
+    context.buffer->WriteContiguousData(8 + ndigits * 2, [&](auto* data) {
```

Comment:
Why do we use WriteContiguousData instead of GetContiguousData? If we return same size